src/lib/bootstrap_controller/bootstrap_controller.ml

```diff
@@ -0,0 +1,244 @@
+
+module Make (Inputs : Inputs_intf) :
+  Bootstrap_controller_intf
+  with type network := Inputs.Network.t
+   and type transition_frontier := Inputs.Transition_frontier.t
+   and type external_transition := Inputs.External_transition.t
+   and type ancestor_prover := Ancestor.Prover.t = struct
+  open Inputs
+
+  type state_with_root =
+    { state: Consensus.Mechanism.Protocol_state.value
+    ; root: Consensus.Mechanism.Protocol_state.value }
+
+  type t =
+    { syncable_ledger: Syncable_ledger.t
+    ; logger: Logger.t
+    ; ancestor_prover: Ancestor.Prover.t
+    ; mutable best_with_root: state_with_root
+    ; network: Network.t }
+
+  let isn't_worth_getting_root t candidate time_received =
+    match
+      Consensus.Mechanism.select ~logger:t.logger
+        ~existing:
+          (Consensus.Mechanism.Protocol_state.consensus_state
+             t.best_with_root.state)
+        ~candidate:
+          (Consensus.Mechanism.Protocol_state.consensus_state candidate)
+        ~time_received
+    with
+    | `Keep -> true
+    | `Take -> false
+
+  let received_bad_proof t e =
+    (* TODO: Punish *)
+    Logger.faulty_peer t.logger !"Bad ancestor proof: %{sexp:Error.t}" e
+
+  let done_syncing_root t =
+    Option.is_some (Syncable_ledger.peek_valid_tree t.syncable_ledger)
+
+  let length protocol_state =
+    Consensus.Mechanism.Protocol_state.consensus_state protocol_state
+    |> Consensus.Mechanism.Consensus_state.length |> Coda_numbers.Length.to_int
+
+  let get_ancestor ~network ({Ancestor.Input.descendant; generations} as input)
+      =
+    let peers = Network.random_peers network 8 in
+    let open Deferred.Or_error.Let_syntax in
+    Deferred.Or_error.find_map_ok peers ~f:(fun peer ->
+        match%bind
+          Network.prove_ancestry network peer (descendant, generations)
+        with
+        | Some proof -> Deferred.Or_error.return proof
+        | None ->
+            Deferred.Or_error.errorf
+              !"Peer %{sexp:Kademlia.Peer.t} does not have proof for \
+                %{sexp:Ancestor.Input.t}"
+              peer input )
+
+  let on_transition t ~ledger_hash_table (transition, time_received) =
+    let module Ps = Consensus.Mechanism.Protocol_state in
+    let candidate = External_transition.protocol_state transition in
+    if isn't_worth_getting_root t candidate time_received then Deferred.unit
+    else
+      let previous_state_hash = Ps.previous_state_hash candidate in
+      (* TODO: This may be an off by one *)
+      let input : Ancestor.Input.t =
+        { descendant= previous_state_hash
+        ; generations= length t.best_with_root.root }
+      in
+      let rec get_root () =
+        let%bind res = get_ancestor ~network:t.network input in
+        if
+          done_syncing_root t
+          || isn't_worth_getting_root t candidate time_received
+        then Deferred.unit
+        else
+          match res with
+          | Error e ->
+              Logger.error t.logger !"%{sexp:Error.t}" e ;
```

Comment:
can you log something more useful here, what is this error and why is it happening?